### PR TITLE
HCLANG-64: Explained output language and kotlin

### PR DESCRIPTION
**Pull request by C0DK**
### Description
added a new section, where we explain why we choose kotlin as well as c++ as output language.

### Changelog
 - Added section regarding languages of the compiler

### Issues solved 
add these with the respective key as HCLANG-1 (as an example)
 - HCLANG-64

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding code [if applicable]
- Check [coding conventions](https://kotlinlang.org/docs/reference/coding-conventions.html)
- Are public methods documented? 
- Are there testcases for new features / methods?
- Have you tried running the source and tested things that cannot be test-cased?

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: report/4.Solution/LanuageOfCompiler.tex [hunk 1]
```diff
@@ -0,0 +1,30 @@
+\section{The languages of the compiler}
+When creating a compiler there are three languages of relevance, as see in the tombstone diagram\ref{fig:TStoneExample}.
```
Comment (jones1994) [inline report/4.Solution/LanuageOfCompiler.tex]: Change "see" -> "seen"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+the input language, in the case of this report HCL, the output language, which is going to be C++. 
+
+\begin{figure}[H]
+	\centering
+	\includegraphics[width=\textwidth/2+\textwidth/4]{4.Solution/images/T-diagram.png}
+	\caption{
+		Example of T-stone diagram showcasing an example of three languages for a given compiler.\cite{TStoneWiki}
```
Comment (jones1994) [inline report/4.Solution/LanuageOfCompiler.tex]: Change "T-stone" -> "Tomb stone diagram"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	}
+	\label{fig:TStoneExample}
+\end{figure}
+\subsection{Language of the compiler}
+The compiler was written in Kotlin\cite{KotlinWebsite}, a Java based language.
+This was based on a number of factors.
+The group wanted to try a new language, but didn't have a lot of time, so the language had to look like known paradigms.
```
[the remark this answers is not in the document]
Comment (jones1994) [reply report/4.Solution/LanuageOfCompiler.tex]: Also, if no reformulation is done; Change "didn't" -> "did not"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+\subsection{Language of the compiler}
+The compiler was written in Kotlin\cite{KotlinWebsite}, a Java based language.
+This was based on a number of factors.
+The group wanted to try a new language, but didn't have a lot of time, so the language had to look like known paradigms. 
+The group knew C\# and a bit Java, however as the HCL language has a lot of functional features, it was deemed important that the group became comfortable with a language that, at the very least was multi paradigm.
+Kotlin is java-based and is similar, in syntax, to C\# and java, which was important.
+However it is still possible to write methods that is utilizing a functional domain.
+
+This grants the develop af large amount of possibilities, while still being relatively easy to learn.
+A disadvantage could be that low-level languages might have faster executing time, but as HCL is built with the Arduino system in mind, it is not possible to create vast programs, as the architecture has some hardware limitations, so execution time of the compiler will never really affect the end user, as the difference shouldn't be noticeable.
```
Comment (jones1994) [inline report/4.Solution/LanuageOfCompiler.tex]: Maybe change "Arduino system" -> "Arduino architecture", since it is properly defined.

Comment (jones1994) [reply report/4.Solution/LanuageOfCompiler.tex]: Also change "shouldn't" -> "should not"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+\subsection{Output language of the compiler}
+The group initially wanted to compile to assembly, primarily for educational purposes, however as the initial brainstorming took place, the group quickly realize that they wanted a big set of possibilities, and that would take a lot of development time simply doing code generation. 
+After some discussion it was agreed upon, to compile to c++, the language that developers normally write for the Arduino, to save time.
+When writing code in the Arduino IDE, it gives access to a couple of predefined methods for interacting with I/O devices, and then through a c/c++ compiler. \cite{ArFAQ}
```
[the remark this answers is not in the document]
Comment (jones1994) [reply report/4.Solution/LanuageOfCompiler.tex]: The \cite needs to be before the dot.

File: report/4.Solution/LanuageOfCompiler.tex [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+In the case of the HCL compiler the three languages are as follows:
+
+\begin{align*}
+input \to & HCL \\
+output \to & C++/C \\
+compiler \to & Kotlin \\
+\end{align*}
+
+\subsection{Language of the compiler}
+The compiler was written in Kotlin\cite{KotlinWebsite}, a Java based language.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: I am not a fan of "A java based language". Try something more along the lines of: "which is a relatively new language that runs on the java virtual machine.
It also uses some of Java's standard libraries and it kan operate with Java code"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+This was based on a number of factors.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: "Kotlin was chosen for a number of reasons" Think this is better.

File: report/4.Solution/LanuageOfCompiler.tex [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+The group wanted to experiment with using a language they had no prior experience with for the duration of the project, but because of time limitations, a new language had to adhere to already known paradigms.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: Change first part to: "The group wanted to try a language"
Also (and this is probably useless): "I don't like the they had no prior experience", but don't know what to change it to

File: report/4.Solution/LanuageOfCompiler.tex [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+The group knew C\# and a bit of Java, however as the HCL language has a lot of functional features, it was deemed important that the group became comfortable with a language that, at the very least was multi paradigm.
+Kotlin is java-based and is similar, in syntax, to C\# and java, which made it easier to learn from the perspective of the group.
+However it is still possible to write methods that is utilizing a functional domain.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: "methods that utilizes a functional domain"

File: report/4.Solution/LanuageOfCompiler.tex [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+This grants the develop af large amount of possibilities, while still being relatively easy to learn.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: This sentence:
Potentially:
"This allows the developer to decide how he/she wants to develop, while still being easy to learn"

Comment (nillas12) [reply report/4.Solution/LanuageOfCompiler.tex]: This sentence was on the first part of it, it sounds strange, Sorry.

File: report/4.Solution/LanuageOfCompiler.tex [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+A disadvantage could be that low-level languages might have faster executing time, but as HCL is built with the Arduino system in mind, it is not possible to create vast programs, as the architecture has some hardware limitations, so execution time of the compiler will never really affect the end user, as the difference should not be noticeable.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: Perhaps look at splitting this sentence into two. In particular around the first , Potentially: ("execution time. However, as HCL is....")

File: report/4.Solution/LanuageOfCompiler.tex [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+\subsection{Output language of the compiler}
+The group initially wanted to compile to assembly, primarily for educational purposes, however as the initial brainstorming took place, the group quickly realized that they wanted a big set of possibilities, which would waste a lot of development time on code generation. 
+After some discussion the group agreed upon compiling to C++, the language that developers normally write for the Arduino, inorder to save time.
```
Comment (nillas12) [inline report/4.Solution/LanuageOfCompiler.tex]: upon -> on
inorder -> in order
Also I don't like the part after the first comma: Potentially: "the language that software for the arduino is normally written in, ..."

File: report/4.Solution/LanuageOfCompiler.tex (hunk 6 above)
Comment (nillas12) [reply report/4.Solution/LanuageOfCompiler.tex]: I can see I am an idiot: Kan to can

**Comment by technight** [comment]
I fixed the java based part of the sentence. I can see that this hided a lot of @nillas12 comments. I think you should show outdated functions by him in order to fix them.

**Pull request merged**